src/Kestrel/Internal/ConfigurationReader.cs

```diff
@@ -0,0 +1,136 @@
+        }
+
+        public IDictionary<string, CertificateConfig> Certificates
+        {
+            get
+            {
+                if (_certificates == null)
+                {
+                    ReadCertificates();
+                }
+
+                return _certificates;
+            }
+        }
+
+        public IEnumerable<EndpointConfig> Endpoints
+        {
+            get
+            {
+                if (_endpoints == null)
+                {
+                    ReadEndpoints();
+                }
+
+                return _endpoints;
+            }
+        }
+
+        private void ReadCertificates()
+        {
+            _certificates = new Dictionary<string, CertificateConfig>(0);
+
+            var certificatesConfig = _configuration.GetSection("Certificates").GetChildren();
+            foreach (var certificateConfig in certificatesConfig)
+            {
+                _certificates.Add(certificateConfig.Key, new CertificateConfig(certificateConfig));
+            }
+        }
+
+        private void ReadEndpoints()
+        {
+            _endpoints = new List<EndpointConfig>();
+
+            var endpointsConfig = _configuration.GetSection("Endpoints").GetChildren();
+            foreach (var endpointConfig in endpointsConfig)
+            {
+                // "EndpointName": {
+                //    "Url": "https://*:5463",
+                //    "Certificate": {
+                //        "Path": "testCert.pfx",
+                //        "Password": "testPassword"
+                //    }
+                // }
+                
+                var url = endpointConfig["Url"];
+                if (string.IsNullOrEmpty(url))
+                {
+                    throw new InvalidOperationException(KestrelStrings.FormatEndpointMissingUrl(endpointConfig.Key));
+                }
+
+                var endpoint = new EndpointConfig()
+                {
+                    Name = endpointConfig.Key,
+                    Url = url,
+                    ConfigSection = endpointConfig,
+                    Certificate = new CertificateConfig(endpointConfig.GetSection("Certificate")),
+                };
+                _endpoints.Add(endpoint);
+            }
+        }
+    }
+
+    internal class EndpointConfig
+    {
+        public string Name { get; set; }
+        public string Url { get; set; }
+        public IConfigurationSection ConfigSection { get; set; }
+        public CertificateConfig Certificate { get; set; }
+    }
+
+    // "CertificateName": {
+    //      "Path": "testCert.pfx",
+    //      "Password": "testPassword"
+    // }
+    internal class CertificateConfig
+    {
+        public CertificateConfig(IConfigurationSection configSection)
+        {
+            ConfigSection = configSection;
+        }
+
+        public IConfigurationSection ConfigSection { get; }
+
+        public bool Exists => ConfigSection.GetChildren().Any();
```

Comment:
Nit: you could consider using the Exists extension method which is similar.
 https://github.com/aspnet/Configuration/blob/dev/src/Config.Abstractions/ConfigurationExtensions.cs#L76

You can consider using section.Bind(this) if you wanted to avoid all the ConfigSection property lookups where the name is the property name.